### PR TITLE
Support out arguments in Variable trigonometery operations

### PR DESCRIPTION
**Pull request by DanNixon**
- Support out arguments on `Variable` trig operations
- Add missing unit tests for Python exports
- Add a benchmark for `Variable::sin` (I used this to make sure my changes to the existing implementation were not slower and have just left it in place)

Re #815.

### CODE REVIEW

File: core/trigonometry.cpp [hunk 1]
```diff
@@ -11,56 +11,170 @@
...
+    makeVariable<double>(Dims(), Shape(), units::Unit(units::rad / units::deg),
+                         Values{pi<double> / 180.0});
 
-Variable sin(const Variable &var) {
+Variable sin(const VariableConstProxy &var) {
   using std::sin;
-  const Variable &rad = var.unit() == units::deg ? to_rad(var) : var;
+  Variable out(var);
+  if (var.unit() == units::deg)
+    out *= deg_to_rad;
```
Comment (SimonHeybrock) [inline core/trigonometry.cpp]: This is forcing an extra copy now, so this should be slower than before?

Comment (DanNixon) [reply core/trigonometry.cpp]: Yes. I'm struggling to find a clean way to handle this that does not end up with more code duplication than I'd like.

The thought did occur to me though that should this even do a unit conversion if we are following the principle of small isolated operations?

Comment (SimonHeybrock) [reply core/trigonometry.cpp]: I see your point about not doing a unit conversion. In this case, however, this is generic enough to be justified. No one would be surprise of `sin(90 deg)` works.

Regarding the copy, better (but not ideal):
1. Create `out` as done in your changes.
2. Forward to call with out arg. This avoids the second copy, and should thus be much closer to optimal performance.

We still end up *iterating* the array twice (copy, then apply), even if no copy is required, but iteration should be considerably cheaper than copies involving memory allocation.

File: core/trigonometry.cpp [hunk 2]
```diff
@@ -11,56 +11,161 @@
...
+  using std::sin;
+  out.assign(var);
+  if (var.unit() == units::deg)
+    out *= deg_to_rad;
+  transform_in_place<pair_self_t<double, float>>(
+      out, out,
+      overloaded{transform_flags::expect_no_variance_arg<0>,
+                 transform_flags::expect_no_variance_arg<1>,
+                 [](auto &x, const auto &y) { x = sin(y); }});
+  return out;
```
Comment (SimonHeybrock) [inline core/trigonometry.cpp]: This ends up iterating over the array up to *3* times. The way it is written you also could have avoided the `transform` with 2 arguments and just use 1 arg instead.

That being said, for `out=in` with input in `rad` there is only a single iteration, so not everything is inefficient. I suggest you just switch back to a single-arg transform and leave the remaining possible "optimizations" until we know it is performance critical.

Comment (DanNixon) [reply core/trigonometry.cpp]: > This ends up iterating over the array up to 3 times.

I don't think there is a nice way around this if you want to keep the unit conversion.
I did try and move the conversion into the transform but Boost units makes this difficult.

Comment (DanNixon) [reply core/trigonometry.cpp]: Even so, the worst theoretical case (non in place with input in degress) is still faster then it originally was with these changes.

Comment (SimonHeybrock) [reply core/trigonometry.cpp]: Yes, as I said: Leave it, just fix the `transform_in_place`.

Comment (SimonHeybrock) [reply core/trigonometry.cpp]: (I see you did, github notifications still do not work with force-pushes, apparently)

**Pull request merged**